doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+
+#### Scope
+
+The value of `self` inside the question node definition block (but outside other blocks) is the instance of the relevant question node. Thus in the example above, the calls to `#option` and `#next_node` are made on the instance of the question node.
+
+In the same way that there is a single instance of each flow class, there is only ever a single instance of each question definition in the system at any one time i.e. a flow instance has an invariant set of question node definition instances.
+
+Since the same instance of a question definition is used across multiple requests, it's important that no request-specific state is stored on them.
+
+The value of `self` inside the `next_node` blocks (or the other optional blocks) is an instance of `SmartAnswer::State` ([see below](#state)), i.e. *not* an instance of a question node.
+
+```ruby
+def define
+  # self = instance of SmartAnswer::Flow
+
+  value_question :question_key do
+    # self = instance of SmartAnswer::Question::Value
+
+    on_response do |response|
+      # self = instance of SmartAnswer::State
+    end
+
+    next_node do
+      # self = instance of SmartAnswer::State
+    end
+  end
+end
+```
+
+#### Execution
+
+The code inside the question node definition block (but outside other blocks) is executed at _flow definition time_, not at request time. However, the code inside the `next_node` blocks (or the other optional blocks) is executed at _request time_, not at flow definition time.
+
+```ruby
+def define
+  # executed at flow definition time
+
+  value_question :question_key do
+    # executed at flow definition time
+
+    on_response do |response|
+      # executed at request time
+    end
+
+    next_node do
+      # executed at request time
+    end
+  end
+end
+```
+
+#### State
+
+The state object is intended to store all request-specific state, keeping that away from the instance of the flow which is reused across multiple requests. `SmartAnswer::State` inherits from `OpenStruct` and uses [`BasicObject#method_missing`][method-missing]
+to allow arbitrary "state variables" to be written and read.
+
+Since the request path only includes the user's responses and *not* the question keys, and the app is stateless, *every* request has to be processed by walking through the question definition nodes starting at the first one. As a request is processed, the state is duplicated using [`Object#dup`][object-dup] in each "transition" to a new node.
+
+It's important to note that `Object#dup` does not do a "deep" copy. Thus any "state variables" set on the state which are references to other objects will continue to reference the same instances of those other object - those objects will *not* themselves be duplicated. The *only* exceptions to this are two built-in state variables, `responses` & `path` ([see below](#built-in-state-variables)) which are themselves duplicated using `Object#dup` in `State#initialize_copy`.
```

Comment:
I've added three inline code blocks to show some examples of how this works. See fix-up commit: 0ea902b3081700211ac1de61ad12139b5f74c530.